olmo/config.py

```diff
@@ -457,6 +462,61 @@ class ModelConfig(BaseConfig):
     See :data:`TrainConfig.precision` instead.
     """
 
+    moe_num_experts: Optional[int] = 8
+    """
+    The number of experts to use in the MoE block.
+    """
+
+    moe_top_k: Optional[int] = 2
```

Comment:
If these are `Optional`, what does it mean when it's `None`?

Reply:
They're optional when no MoE is used, otherwise required. Is this not an acceptable usage of `Optional[int]`? Can change it

Reply:
I can change it to option 1) if others agree? Note that there's other params not following this:
```
    embedding_size: Optional[int] = 50304
    gen1_gc_interval: Optional[int] = 1
    distributed_strategy: Optional[DistributedStrategy] = DistributedStrategy.fsdp
    fsdp: Optional[FSDPConfig] = field(default_factory=FSDPConfig)
    auxiliary_loss_multiplier: Optional[float] = 1e-4
```

Reply:
Do you actually rely on the defaults you put in here anywhere? If not, let's go with Shane's version, and default these to `None`. I assume something somewhere will fail if they are not set and you need them.

Reply:
> Do you actually rely on the defaults you put in here anywhere?

Yes quite a lot, e.g. the loss weights; the use of dropless MoEs (moe_dropless); leaving moe_interleave,moe_lbl_in_fp32,moe_shared_expert as False

Actually, I don't think setting them all to None is a good idea, as it means that everytime we add a new MoE-specific configuration parameter all MoE configs become outdated since every MoE-specific configuration parameter is Optional in that dense.

I can also remove the `Optional` from it as they have defaults anyways but then as seen in the examples I pasted above, we do have `Optional` config params with default values in the codebase anyways.

Reply:
If it doesn't break everything, I'd prefer to have a special config object for MoE, which is `Optional`, but none of the items inside of that object are `Optional`. This may break backwards compatibility with the model we already released though?

Reply:
Yes it would break compat with the configs we released but can pin a commit to our released repo if people want to reuse our configs to reproduce things exactly

Reply:
Hm, that's unfortunate, but I think I prefer the `MoEConfigObject`. It reduces the impact on old-school dense model training.

Reply:
I guess it would make the name ModelConfig a bit inaccurate though; maybe it should inherit from ModelConfig or sth